fastly/cloudfiles.go

```diff
@@ -0,0 +1,255 @@
+	Path              string `form:"path,omitempty"`
+	Region            string `form:"region,omitempty"`
+	Placement         string `form:"placement,omitempty"`
+	Period            uint   `form:"period,omitempty"`
+	GzipLevel         uint   `form:"gzip_level,omitempty"`
+	Format            string `form:"format,omitempty"`
+	FormatVersion     uint   `form:"format_version,omitempty"`
+	ResponseCondition string `form:"response_condition,omitempty"`
+	MessageType       string `form:"message_type,omitempty"`
+	TimestampFormat   string `form:"timestamp_format,omitempty"`
+	PublicKey         string `form:"public_key,omitempty"`
+}
+
+// CreateCloudfiles creates a new Fastly Cloudfiles.
+func (c *Client) CreateCloudfiles(i *CreateCloudfilesInput) (*Cloudfiles, error) {
+	if i.Service == "" {
+		return nil, ErrMissingService
+	}
+
+	if i.Version == 0 {
+		return nil, ErrMissingVersion
+	}
+
+	path := fmt.Sprintf("/service/%s/version/%d/logging/cloudfiles", i.Service, i.Version)
+	resp, err := c.PostForm(path, i, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var cloudfiles *Cloudfiles
+	if err := decodeJSON(&cloudfiles, resp.Body); err != nil {
+		return nil, err
+	}
+	return cloudfiles, nil
+}
+
+// GetCloudfilesInput is used as input to the GetCloudfiles function.
+type GetCloudfilesInput struct {
+	// Service is the ID of the service. Version is the specific configuration
+	// version. Both fields are required.
+	Service string
+	Version int
+
+	// Name is the name of the Cloudfiles to fetch.
+	Name string
+}
+
+// GetCloudfiles gets the Cloudfiles configuration with the given parameters.
+func (c *Client) GetCloudfiles(i *GetCloudfilesInput) (*Cloudfiles, error) {
+	if i.Service == "" {
+		return nil, ErrMissingService
+	}
+
+	if i.Version == 0 {
+		return nil, ErrMissingVersion
+	}
+
+	if i.Name == "" {
+		return nil, ErrMissingName
+	}
+
+	path := fmt.Sprintf("/service/%s/version/%d/logging/cloudfiles/%s", i.Service, i.Version, url.PathEscape(i.Name))
+	resp, err := c.Get(path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var cloudfiles *Cloudfiles
+	if err := decodeJSON(&cloudfiles, resp.Body); err != nil {
+		return nil, err
+	}
+	return cloudfiles, nil
+}
+
+// UpdateCloudfilesInput is used as input to the UpdateCloudfiles function.
+type UpdateCloudfilesInput struct {
+	// Service is the ID of the service. Version is the specific configuration
+	// version. Both fields are required.
+	Service string
+	Version int
+
+	// Name is the name of the Cloudfiles to update.
+	Name string
+
+	NewName           string `form:"name,omitempty"`
+	AccessKey         string `form:"access_key,omitempty"`
+	BucketName        string `form:"bucket_name,omitempty"`
+	Path              string `form:"path,omitempty"`
+	Region            string `form:"region,omitempty"`
+	Placement         string `form:"placement,omitempty"`
+	Period            uint   `form:"period,omitempty"`
+	GzipLevel         uint   `form:"gzip_level,omitempty"`
```

Comment:
We should discuss how we want to handle scalar values and omitempty in the update input structs. We've recently started to move other update structs over to pointer values. The problem is the combination of Go's type zero values, JSON marshalling with the omitempty struct field tag, and a users intent to actually set a zero value. This blog post does a much better job at explaining the problem and well worth a read: https://willnorris.com/2014/05/go-rest-apis-and-pointers/ .

Therefore this current implementation would make it impossible to update these values to their zero values i.e. 0 in this case. A more in-depth explanation and discussion of recent decisions made in the library can be found in this PR: https://github.com/fastly/go-fastly/pull/142#discussion_r368996195

Reply:
I believe I've addressed this in https://github.com/fastly/go-fastly/pull/162/commits/9c90c20aa2b31ea93689b69aa2f102dd24af9101.